documentation/projects/proposals/trust_and_safety/content_report_moderation/20231208-implementation_plan_django_admin_moderator_access.md

```diff
@@ -0,0 +1,587 @@
+  metadata
+
+#### `ModerationDecision`
+
+We will create a new `ModerationDecision` model for each media type, to
+accompany the existing report models. The decision model has a many-to-one
+relationship with the _media_ and tracks the history of moderation decisions
+made for the media, including explanatory notes from moderators. It is not
+possible to use `LogEntry`, like the project plan suggests to do, because
+`LogEntry`'s `object_id` column is not indexed, and querying it to find all the
+changes to all content reports for a given work would not scale.
+
+This table anticipates additional abilities for moderators in the future
+including reversing previous decisions (currently impossible, but needed if
+users and especially creators can contest decisions) or explicitly marking works
+as "not sensitive" (for example, if the work was only marked sensitive because
+of erroneous sensitive text detection). Right now the data model and approach to
+sensitivity and deindexing don't make these possible. It is outside the scope of
+this project to add those features, but this new table, especially its `action`
+column, is intentionally flexible to allow for these kinds of things in the
+future. This is also why it isn't just a new `explanation` column on the content
+report models: because we want to preserve a _record_ of moderation decisions in
+chronological order, and describes the progressive changes of a work's
+sensitivity designation (or de-indexing). It also separates the moderation
+decision (and metadata) from the report, which has its own important metadata
+that should not be confused with the moderation decision metadata.
+
+Additional, the "Bulk moderation actions" implementation plan that is also part
+of this project would have a good use case for this, with separate action types.
+When moderating all works from a particular creator, for example, we wouldn't
+need to generate reports for each of the works and action them, just a single
+`ModerationDecision` for each work with an appropriate action and consequence.
+This avoids need to create "fabricated" reports for bulk moderation actions that
+wouldn't have reports in the first place. _This is just a suggested additional
+use case and is not a specific implementation detail for that plan. The actual
+plan can and probably will have additional considerations beyond this._
+
+It will have the following columns:
+
+- `id` - basic auto-incrementing id
+- `created_on` - the date the moderation decision was created, in other words,
+  the date the decision was made
+  - Provided by `OpenLedgerModel`
+- `moderator_id` - a foreign key relationship with moderator who made the
+  decision
+- `media_identifier` - a one-to-one relationship to the media the decision is
+  related to
+  - Because deindexing media removes them from the API database, this needs to
+    be a non-constrained column, so not a foreign key, just an indexed
+    _non-unique_ UUID column
```

Comment:
We _do_ actually [delete the media](https://github.com/WordPress/openverse/blob/main/api/api/models/media.py#L329) from the database, in addition to removing it from Elasticsearch. They are not deleted in the _catalog_ database, but we have logic in the ingestion server that prevents them from being added back to the API db during a data refresh as well.

Reply:
It deletes the media as well as removing the ES result: https://github.com/WordPress/openverse/blob/aa16d4f1be7607b12c428886b9890bdd947cc71c/api/api/models/media.py#L329

It's a hard delete from the API database and future reindex jobs exclude them altogether based on the deleted media table. There is also the issue of the recreate causing foreign key issues, but we solve that with Django level contraints elsewhere (`db_constraint=False` [on the report models, for example](https://github.com/WordPress/openverse/blob/eb0e906e7300e32fae945eb9222ed47a50fb72a2/api/api/models/image.py#L111-L119)).

So it's both :slightly_smiling_face: 

Also occurring to me that we might want to make this a many-to-many for the bulk moderation actions in the future :thinking:

Reply:
Oops, sorry, Staci's comment hadn't loaded for me yet when I wrote mine! Staci's got it :+1:

Reply:
Ok, it seems we understand kind of the same here! A deletion registry is still _kept in the API DB_, not in the `<media>` table _but_ in the `api_<media>deleted` or `nsfw_reports*` (oh lord, forgive the names!) tables, depending on the action selected. Otherwise, it wouldn't be possible to determine which works have been marked as sensitive or deleted when reingesting.

My fear was that saying that it's a "_hard delete from the API database_" is not totally correct and might be missing the context of the data refresh process, but if we all understand the same then it's fine.

Reply:
Yup, you've got it. The "hard delete" is from the perspective of just the media object, which is what this identifier is for. I left out further details for no real reason other than they felt unrelated, but I can see the potential confusion if someone doesn't have all the domain context... but then I don't know when to stop writing :joy: